File: src/main/java/marquez/service/models/Dataset.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package marquez.service.models;

import com.google.common.collect.ImmutableList;
import java.time.Instant;
import java.util.List;
import java.util.Optional;
import javax.annotation.Nullable;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;
import marquez.common.models.DatasetName;
import marquez.common.models.DatasetType;
import marquez.common.models.Field;
import marquez.common.models.SourceName;

@EqualsAndHashCode
@ToString
public abstract class Dataset {
  @Getter private final DatasetType type;
  @Getter private final DatasetName name;
  @Getter private final DatasetName physicalName;
  @Getter private final Instant createdAt;
  @Getter private final Instant updatedAt;
  @Getter private final SourceName sourceName;
  @Nullable private final List<Field> fields;
  @Nullable private final List<String> tags;
  @Nullable private final Instant lastModified;
  @Nullable private final String description;

  public Dataset(
      @NonNull final DatasetType type,
      @NonNull final DatasetName name,
      @NonNull final DatasetName physicalName,
      @NonNull final Instant createdAt,
      @NonNull final Instant updatedAt,
      @NonNull final SourceName sourceName,
      @Nullable final List<Field> fields,
      @Nullable final List<String> tags,
      @Nullable final Instant lastModified,
      @Nullable final String description) {
    this.type = type;
    this.name = name;
    this.physicalName = physicalName;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
    this.sourceName = sourceName;
    this.fields = fields;
    this.tags = tags;
    this.lastModified = lastModified;
    this.description = description;
  }

  public List<Field> getFields() {
    return (fields == null) ? ImmutableList.of() : ImmutableList.copyOf(fields);
  }

  public List<String> getTags() {
    return (tags == null) ? ImmutableList.of() : ImmutableList.copyOf(tags);
  }

  public Optional<Instant> getLastModified() {
    return Optional.ofNullable(lastModified);
  }

  public Optional<String> getDescription() {
    return Optional.ofNullable(description);
  }
}

File: integrations/common/marquez/dataset.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List, Optional, Dict

from marquez.models import DbTableSchema, DbColumn
from openlineage.facet import BaseFacet, DataSourceDatasetFacet, DocumentationDatasetFacet, \
    SchemaDatasetFacet, SchemaField
from openlineage.run import Dataset as OpenLineageDataset


class Source:
    def __init__(
            self,
            scheme: Optional[str] = None,
            authority: Optional[str] = None,
            connection_url: Optional[str] = None,
            name: Optional[str] = None
    ):
        self.scheme = scheme
        self.authority = authority
        self._name = name
        self.connection_url = connection_url

        if (scheme or authority) and name:
            raise RuntimeError('scheme + authority and namespace are exclusive options')

    def __eq__(self, other):
        return self.name == other.name and \
               self.connection_url == other.connection_url

    def __repr__(self):
        authority = '//' + self.authority if self.authority else ''
        return f"Source({self.scheme!r}:{authority} - {self.connection_url!r})"

    @property
    def name(self) -> str:
        if self._name:
            return self._name
        if self.authority:
            return f'{self.scheme}://{self.authority}'
        return f'{self.scheme}'


class Field:
    def __init__(self, name: str, type: str,
                 tags: List[str] = None, description: str = None):
        self.name = name
        self.type = type
        self.tags = tags
        self.description = description

        if self.tags is None:
            self.tags = []

    @staticmethod
    def from_column(column: DbColumn):
        return Field(
            name=column.name,
            type=column.type,
            description=column.description
        )

    def __eq__(self, other):
        return self.name == other.name and \
               self.type == other.type and \
               self.tags == other.tags and \
               self.description == other.description

    def __repr__(self):
        return f"Field({self.name!r},{self.type!r}, \
                       {self.tags!r},{self.description!r})"


class Dataset:
    def __init__(
            self,
            source: Source,
            name: str, fields: List[Field] = None,
            description: Optional[str] = None,
            custom_facets: Dict[str, BaseFacet] = None
    ):
        if fields is None:
            fields = []
        if custom_facets is None:
            custom_facets = {}
        self.source = source
        self.name = name
        self.fields = fields
        self.description = description
        self.custom_facets = custom_facets

    @staticmethod
    def from_table(source: Source,
                   table_name: str,
                   schema_name: str = None,
                   database_name: str = None):
        return Dataset(
            name=Dataset._to_name(
                schema_name=schema_name,
                table_name=table_name,
                database_name=database_name
            ),
            source=source
        )

    @staticmethod
    def from_table_schema(
            source: Source,
            table_schema: DbTableSchema,
            database_name: str = None
    ):
        return Dataset(
            name=Dataset._to_name(
                schema_name=table_schema.schema_name,
                table_name=table_schema.table_name.name,
                database_name=database_name
            ),
            source=source,
            fields=[
                # We want to maintain column order using ordinal position.
                Field.from_column(column) for column in sorted(
                    table_schema.columns, key=lambda x: x.ordinal_position
                )
            ]
        )

    @staticmethod
    def _to_name(table_name: str, schema_name: str = None, database_name: str = None):
        # Prefix the table name with database and schema name using
        # the format: {database_name}.{table_schema}.{table_name}.
        name = [table_name]
        if schema_name is not None:
            name = [schema_name] + name
        if database_name is not None:
            name = [database_name] + name

        return ".".join(name)

    def __eq__(self, other):
        return self.source == other.source and \
               self.name == other.name and \
               self.fields == other.fields and \
               self.description == other.description

    def __repr__(self):
        return f"Dataset({self.source!r},{self.name!r}, \
                         {self.fields!r},{self.description!r})"

    def to_openlineage_dataset(self) -> OpenLineageDataset:
        facets = {
            "dataSource": DataSourceDatasetFacet(
                self.source.name,
                self.source.connection_url
            )
        }
        if self.description:
            facets.update({
                "documentation": DocumentationDatasetFacet(
                    description=self.description
                )
            })

        if self.fields is not None and len(self.fields):
            facets.update({
                "schema": SchemaDatasetFacet(
                    fields=[
                        SchemaField(field.name, field.type, field.description)
                        for field in self.fields
                    ]
                )
            })

        if self.custom_facets:
            facets.update(self.custom_facets)

        return OpenLineageDataset(
            namespace=self.source.name,
            name=self.name,
            facets=facets
        )
